internal/impl/pure/scanner_json.go

```diff
@@ -0,0 +1,94 @@
+package pure
+
+import (
+	"context"
+	json "encoding/json"
+	"errors"
+	"io"
+
+	"github.com/benthosdev/benthos/v4/public/service"
+)
+
+const (
+	sjsonFieldContinueOnError = "continue_on_error"
+)
+
+func jsonDocumentScannerSpec() *service.ConfigSpec {
+	return service.NewConfigSpec().
+		Stable().
+		Summary("Consumes JSON documents line-by-line.").
+		Fields(
+			service.NewBoolField(sjsonFieldContinueOnError).
+				Description("If a JSON decoding fails due to any error emit an empty message marked with the error and then continue consuming subsequent documents when possible.").
+				Default(false),
+		)
+}
+
+func init() {
+	err := service.RegisterBatchScannerCreator("json", jsonDocumentScannerSpec(),
+		func(conf *service.ParsedConfig, mgr *service.Resources) (service.BatchScannerCreator, error) {
+			return jsonDocumentScannerFromParsed(conf)
+		})
+	if err != nil {
+		panic(err)
+	}
+}
+
+func jsonDocumentScannerFromParsed(conf *service.ParsedConfig) (l *jsonDocumentScannerCreator, err error) {
+	l = &jsonDocumentScannerCreator{}
+	if l.continueOnError, err = conf.FieldBool(sjsonFieldContinueOnError); err != nil {
+		return
+	}
+	return
+}
+
+type jsonDocumentScannerCreator struct {
+	continueOnError bool
+}
+
+func (js *jsonDocumentScannerCreator) Create(rdr io.ReadCloser, aFn service.AckFunc, details *service.ScannerSourceDetails) (service.BatchScanner, error) {
+	return service.AutoAggregateBatchScannerAcks(&jsonDocumentScanner{
+		d:               json.NewDecoder(rdr),
+		r:               rdr,
+		continueOnError: js.continueOnError,
+	}, aFn), nil
+}
+
+func (js *jsonDocumentScannerCreator) Close(context.Context) error {
+	return nil
+}
+
+type jsonDocumentScanner struct {
+	d *json.Decoder
+	r io.ReadCloser
+
+	continueOnError bool
+}
+
+func (js *jsonDocumentScanner) NextBatch(ctx context.Context) (service.MessageBatch, error) {
+	if js.r == nil {
+		return nil, io.EOF
+	}
+
+	msg := service.NewMessage(nil)
+
+	var jsonDocObj any
+
+	if err := js.d.Decode(&jsonDocObj); err != nil {
+		if errors.Is(err, io.EOF) || !js.continueOnError {
+			return nil, err
+		}
```

Comment:
In order to match up with the docs here we should be annotating `msg` with the error returned here.

Reply:
Addressed all of these in 0a167a8